lib/vector-tap/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+            output_patterns.clone(),
+            input_patterns.clone(),
+            format,
+            limit,
+            interval,
+        );
+    }
+
+    let start_time = Instant::now();
+    let stream_duration =
+        duration_ms
+            .map(Duration::from_millis)
+            .unwrap_or(Duration::MAX);
+
+    // Loop over the returned results, printing out tap events.
+    #[allow(clippy::print_stdout)]
+    #[allow(clippy::print_stderr)]
+    loop {
+        let time_elapsed = start_time.elapsed();
+        if time_elapsed >= stream_duration {
+            return Ok(());
+        }
+
+        let message = timeout(stream_duration - time_elapsed, stream.next()).await;
+        match message {
+            Ok(Some(Some(res))) => {
+                if let Some(d) = res.data {
+                    match &output_channel {
+                        OutputChannel::Stdout(formatter) => {
+                            for tap_event in d.output_events_by_component_id_patterns.iter() {
+                                match tap_event {
+                                    GraphQLTapOutputEvents::Log(ev) => {
+                                        println!("{}", formatter.format(ev.component_id.as_ref(), ev.component_kind.as_ref(), ev.component_type.as_ref(), ev.string.as_ref()));
+                                    }
+                                    GraphQLTapOutputEvents::Metric(ev) => {
+                                        println!("{}", formatter.format(ev.component_id.as_ref(), ev.component_kind.as_ref(), ev.component_type.as_ref(), ev.string.as_ref()));
+                                    }
+                                    GraphQLTapOutputEvents::Trace(ev) => {
+                                        println!("{}", formatter.format(ev.component_id.as_ref(), ev.component_kind.as_ref(), ev.component_type.as_ref(), ev.string.as_ref()));
+                                    }
+                                    GraphQLTapOutputEvents::EventNotification(ev) => {
+                                        if !quiet {
+                                            eprintln!("{}", ev.message);
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                        OutputChannel::AsyncChannel (sender_tx) => {
+                            if sender_tx.send(d.output_events_by_component_id_patterns).await.is_err() {
+                                debug!("Could not send events");
+                            }
+                        }
+                    }
+                }
+            }
+            Err(_) =>
+            // If the stream times out, that indicates the duration specified by the user
+            // has elapsed. We should exit gracefully.
+                {
+                    return Ok(())
+                }
+            Ok(_) => return Err(TapExecutorError::NoEventsFound)
```

Comment:
> [!NOTE]  
> This is a bit confusing because we have `Option<Option<...>>`. The first `Option` indicates that there was a conversion error [here](https://github.com/vectordotdev/vector/blob/1579627ee3c63347fa68fe8ab3e0105bbe6ff3d9/lib/vector-api-client/src/subscription.rs#L78) when trying to convert `graphql_client::Response<T::ResponseData>`.